File: packages/python/examples/verify_tron_privy_typed_data.py
"""Verify TRON Privy sign_typed_data by recovering the TRON address
from the signature and comparing to the active wallet address.

Usage:
  AGENT_WALLET_DIR=/tmp/test-wallet \
  AGENT_WALLET_PASSWORD='Abc12345!@' \
  python examples/verify_tron_privy_typed_data.py
"""

from __future__ import annotations

import os

from eth_account.messages import _hash_eip191_message, encode_typed_data
from tronpy import keys

from agent_wallet.core.providers.config_provider import ConfigWalletProvider
from agent_wallet.core.resolver import resolve_wallet_provider

DIR = os.environ.get("AGENT_WALLET_DIR", "/tmp/test-wallet")
PASSWORD = os.environ.get("AGENT_WALLET_PASSWORD", "")
NETWORK = os.environ.get("AGENT_WALLET_NETWORK", "tron")

if not PASSWORD:
    raise SystemExit("AGENT_WALLET_PASSWORD is required to access local_secure wallets.")

provider = resolve_wallet_provider(network=NETWORK, dir=DIR)
if not isinstance(provider, ConfigWalletProvider):
    raise SystemExit("Expected a config-backed provider. Check AGENT_WALLET_DIR.")


def build_typed_data() -> dict:
    return {
        "domain": {
            "name": "AgentWallet",
            "version": "1",
            "chainId": 1,
        },
        "types": {
            "EIP712Domain": [
                {"name": "name", "type": "string"},
                {"name": "version", "type": "string"},
                {"name": "chainId", "type": "uint256"},
            ],
            "Message": [{"name": "contents", "type": "string"}],
        },
        "primaryType": "Message",
        "message": {"contents": "Hello"},
    }


def hash_typed_data(data: dict) -> bytes:
    signable = encode_typed_data(full_message=data)
    return bytes(_hash_eip191_message(signable))


def recover_tron_address(signature_hex: str, digest: bytes) -> str:
    sig_bytes = bytes.fromhex(signature_hex)
    if len(sig_bytes) != 65:
        raise ValueError(f"Expected 65-byte signature, got {len(sig_bytes)}")
    v = sig_bytes[64]
    if v in (0, 1):
        recovery = v
    elif v in (27, 28):
        recovery = v - 27
    else:
        raise ValueError(f"Invalid recovery id: {v}")
    sig = keys.Signature(sig_bytes[:64] + bytes([recovery]))
    pub = sig.recover_public_key_from_msg_hash(digest)
    return pub.to_base58check_address()


async def main() -> None:
    active_id = provider.get_active_id()
    if not active_id:
        raise SystemExit("No active wallet set.")

    wallet = await provider.get_active_wallet(NETWORK)
    address = await wallet.get_address()

    typed_data = build_typed_data()
    signature = await wallet.sign_typed_data(dict(typed_data))
    digest = hash_typed_data(typed_data)
    recovered = recover_tron_address(signature, digest)

    print(f"Active wallet: {active_id}")
    print(f"Wallet address: {address}")
    print(f"Recovered addr: {recovered}")
    print(f"Signature: {signature}")
    print(f"Verified: {recovered == address}")


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())

File: packages/typescript/examples/verify-tron-privy-typed-data.ts
/**
 * Verify TRON Privy signTypedData by recovering the TRON address
 * from the signature and comparing to the active wallet address.
 *
 * Usage:
 *   AGENT_WALLET_DIR=/tmp/test-wallet npx tsx examples/verify-tron-privy-typed-data.ts
 */

import { secp256k1 } from '@noble/curves/secp256k1'
import bs58check from 'bs58check'
import { hashTypedData, keccak256 } from 'viem'

import { ConfigWalletProvider, resolveWalletProvider } from '../src/index.js'
import { requireEip712Wallet } from './example-utils.js'

const DIR = process.env.AGENT_WALLET_DIR ?? '/tmp/test-wallet'
const NETWORK = process.env.AGENT_WALLET_NETWORK ?? 'tron:mainnet'

function resolveConfigProvider(): ConfigWalletProvider {
  const provider = resolveWalletProvider({ dir: DIR, network: NETWORK })
  if (!(provider instanceof ConfigWalletProvider)) {
    throw new Error('Expected a config-backed provider. Check AGENT_WALLET_DIR.')
  }
  return provider
}

const provider = resolveConfigProvider()

function stripHexPrefix(value: string) {
  return value.startsWith('0x') ? value.slice(2) : value
}

function tronAddressFromPublicKey(pubkey: Uint8Array) {
  const uncompressed = pubkey[0] === 4 ? pubkey.slice(1) : pubkey
  const hash = keccak256(uncompressed)
  const addrBytes = Buffer.from(hash.slice(2), 'hex').slice(-20)
  const tronAddrBytes = Buffer.concat([Buffer.from([0x41]), addrBytes])
  return bs58check.encode(tronAddrBytes)
}

function buildTypedData() {
  return {
    domain: {
      name: 'AgentWallet',
      version: '1',
      chainId: 1,
    },
    types: {
      EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'chainId', type: 'uint256' },
      ],
      Message: [{ name: 'contents', type: 'string' }],
    },
    primaryType: 'Message',
    message: {
      contents: 'Hello',
    },
  } as const
}

function cloneTypedData(data: ReturnType<typeof buildTypedData>) {
  return JSON.parse(JSON.stringify(data)) as ReturnType<typeof buildTypedData>
}

function hashTypedDataForTron(data: ReturnType<typeof buildTypedData>) {
  const { domain, types, primaryType, message } = data
  if (!primaryType || !(primaryType in types)) {
    throw new Error(`primaryType must be a key in types: ${primaryType ?? 'undefined'}`)
  }
  return hashTypedData({
    domain: { ...domain, chainId: BigInt(domain.chainId) },
    types,
    primaryType,
    message,
  })
}

async function main() {
  const activeId = provider.getActiveId()
  if (!activeId) {
    throw new Error('No active wallet set.')
  }

  const wallet = requireEip712Wallet(await provider.getActiveWallet(NETWORK))
  const address = await wallet.getAddress()

  const typedData = buildTypedData()
  const signature = await wallet.signTypedData(cloneTypedData(typedData))
  const sigHex = stripHexPrefix(signature)
  const sigBytes = Buffer.from(sigHex, 'hex')
  if (sigBytes.length !== 65) {
    throw new Error(`Expected 65-byte signature, got ${sigBytes.length}`)
  }
  const recovery = sigBytes[64] - 27
  if (recovery !== 0 && recovery !== 1) {
    throw new Error(`Invalid recovery id: ${sigBytes[64]}`)
  }

  const hashHex = hashTypedDataForTron(cloneTypedData(typedData))
  const hashBytes = Buffer.from(hashHex.slice(2), 'hex')
  const sig = secp256k1.Signature.fromCompact(sigBytes.slice(0, 64)).addRecoveryBit(recovery)
  const pub = sig.recoverPublicKey(hashBytes)
  const recovered = tronAddressFromPublicKey(pub.toRawBytes(false))

  console.log(`Active wallet: ${activeId}`)
  console.log(`Wallet address: ${address}`)
  console.log(`Recovered addr: ${recovered}`)
  console.log(`Signature: ${signature}`)
  console.log(`Verified: ${recovered === address}`)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
